### PR TITLE
DBC: Add support for STRING type in GenMsgSendType

### PR DESCRIPTION
**Pull request by Mrc1979**
BUG fixing: 
---------------------
Prevents ValueError in case of  result = message_attributes['GenMsgSendType'].value is a String value.

Necessary in case of Evnet-based sending like in that dbc attribute expression:

BA_ "GenMsgSendType" BO_ 1373 "Event";
BA_ "GenMsgCycleTime" BO_ 1373 0;

in Vector CANdb++ generated dbcs.

Greetings Maurice

### CODE REVIEW

**Comment by Mrc1979** [comment]
Hi all, please approve that request, solving a Vector dbc parsing issue in case of Event based and not Cyclic messages.

**Comment by andlaus** [comment]
The code itself looks fine to me, but unfortunately I've got no idea whether what it does is correct or not.

@juleq: Can you assess that?

**Comment by andlaus** [comment]
@Mrc1979: one thing: could you please add or modify a test case so that this new code is triggered?

**Comment by Mrc1979** [comment]
Hi Andreas, there had been an issue when identifying event-based messages. Is not a big one but it prevents cantools running through properly. With my change a sending type, not based on a numeration is accepted. That is the case if you define an event based sending, not a cyclic one as often applied.

Greetings Maurice

⁣BlueMail for Android herunterladen ​

Am 24. Feb. 2022, 18:15, um 18:15, Andreas Lauser ***@***.***> schrieb:
>The code itself looks fine to me, but unfortunately I've got no idea
>whether what it does is correct or not.
>
***@***.***: Can you assess that?

**Comment by Mrc1979** [comment]
Sorry forgot: Without that line a standard DBC file, e g. created by CANeditor++ of Vector causes by literal mismatch a ValueError. 
If needed I can demonstrate.


⁣BlueMail for Android herunterladen ​

Am 24. Feb. 2022, 18:15, um 18:15, Andreas Lauser ***@***.***> schrieb:
>The code itself looks fine to me, but unfortunately I've got no idea
>whether what it does is correct or not.
>
***@***.***: Can you assess that?

**Comment by Mrc1979** [comment]
Hi Andreas, I'm working in company in which we for measurements defined the DBC files. So for a test case I cannot deliver a DBC. Can you send me and link to testing function for unit Test placed for that package then I try to adapt using one of the cantools dbcs for evaluation purposes. Greetings Maurice

⁣BlueMail for Android herunterladen ​

Am 24. Feb. 2022, 18:16, um 18:16, Andreas Lauser ***@***.***> schrieb:
***@***.***: one thing: could you please add or modify a test case so that

**Comment by andlaus** [comment]
> So for a test case I cannot deliver a DBC.

If you're talking about files that are used in production that's fair enough. (I would be *very* surprised if you could.) 

> Can you send me and link to testing function for unit Test placed for that package then I try to adapt using one of the cantools dbcs for evaluation purposes

How about modifying e.g. [`motohawk.dbc`](https://github.com/cantools/cantools/blob/master/tests/files/dbc/motohawk.dbc). Alternatively you can also  also create a synthetic one from scratch if this is easier. After that, tickle the new code in e.g. one of the [database tests](https://github.com/cantools/cantools/blob/master/tests/test_database.py).

**Comment by Mrc1979** [comment]
Hi Andreas, ok, I will try tomorrow. Greetings Maurice

⁣BlueMail for Android herunterladen ​

**Comment by Mrc1979** [comment]
Cool I checked motohawk.dbc, will be easy to extend for a test case I think 👍

⁣BlueMail for Android herunterladen ​

**Comment by juleq** [comment]
Hello @Mrc1979.

Is there a reason that you do not use the usual way of using the ENUM that Vector predefines (or you can create it yourself in CANdb++) for GenMsgSendType in dbc templates like the Vector_IL one? That way you have the advantage of using predefined strings from a drop down, which is safer then allowing a custom STRING. The type of the attribute is INT then, because it stores the index of the selected ENUM value. Example:

`
BA_DEF_ BO_  "GenMsgSendType" ENUM  "Cyclic","Event","NotUsed","NoMsgSendType";
BA_DEF_DEF_  "GenSigSendType" "Cyclic";
BA_ "GenMsgSendType" BO_ 42 1;
`

Line 1: The BA_DEF_ BO_ is the attribute definition. Its type is ENUM.
Line 2: For some funny reason, Vector themselves use a STRING to define the default value for this definition instead of setting it to 0.
Line 3: This is setting the attributes value to 1 (Event, a custom value that I have added to the ENUM in CANdb++) and references it to a message with id 0x2A.

Note: You can add new and edit existing ENUMs in CANdb++ -> View -> Attribute Definitions. Take case that you select the right one to edit, you can also reference to signals (GenSIGSendType).

The funny reason in line 2 would make me open for the requested change in your PR. Good that you do a test for that. You could also detail your usage of cantools that revealed this error for you by showing us the commands or lines you were using. Kind of like you did initially but a little more context before the code line and the relevant BA_DEF_ and BA_DEF_DEF_ lines of your dbc.

**Comment by amitrevska-dev** [comment]
Hello. I am also facing the same issue. Kindly offer support here. Thank you.

**Comment by Mrc1979** [comment]
@andlaus Hi Andreas,
me and a friend (@vignesRectifier) tested our implementation and extended the given tests by another unit test.
You se in the log that with my small fix a file using the Event-Flag directly created via CANdb++ can be rea in and handled without throwing the mentioned value error.

[UnitTest_Final_Console.txt](https://github.com/cantools/cantools/files/8215812/UnitTest_Final_Console.txt)
[UnitTest_Error_Console.txt](https://github.com/cantools/cantools/files/8215814/UnitTest_Error_Console.txt)

**Comment by Mrc1979** [comment]
I will add the test dbc and test_database.py entension to the PR...

File: cantools/database/can/formats/dbc.py [hunk 1]
```diff
@@ -1386,8 +1386,11 @@ def get_send_type(frame_id_dbc):
 
         try:
             result = message_attributes['GenMsgSendType'].value
-            # Resolve ENUM index to ENUM text
-            result = definitions['GenMsgSendType'].choices[int(result)]
+            
+            # if definitions is enum (otherwise above value is maintained) -> Prevents ValueError
+            if definitions['GenMsgSendType'].choices != None:
```
Comment (andlaus) [inline cantools/database/can/formats/dbc.py:1391]: I'm not 100% sure, but this seems to be more bullet proof as it (probably) works if `definitions['GenMsgSendType'].choices` is defined but does not exhibit an entry for `result`?
```suggestion
            if definitions['GenMsgSendType'].choices is not None and \

                int(result) in definitions['GenMsgSendType'].choices:
```

**Review by andlaus — COMMENTED** [review]
as far as I can tell this is fine, but @juleq is the expert on DBC, so he is the ultimate authority here...

**Comment by juleq** [comment]
Did try @andlaus proposal, but it failed two tests. I do not have time to get into it, was probably the double exception trickery there. Since coverage is good for attributes, I´ll merge.

**Comment by vignesRectifier** [comment]
Thank you for the extension.. now we are eagerly waiting for the next release..

**Comment by Mrc1979** [comment]
Hi Jule, many thanks for fast release. Have a nice weekend. Greetings Maurice

⁣BlueMail for Android herunterladen ​

**Pull request merged**